docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,337 @@
+from dagster import Definitions, job, op
+from dagster._config.structured_config import ConfigurableResource
+
+
+class MyConnectionResource(ConfigurableResource):
+    username: str
+    password: str
+
+    def request(self, endpoint: str) -> Response:
+        return requests.get(
+            f"https://my-api.com/{endpoint}",
+            auth=(self.username, self.password),
+        )
+
+
+# Since MyConnectionResource extends ConfigurableResource, we don't
+# need to wrap it in a Resource[] annotation.
+@op
+def update_service(my_conn: MyConnectionResource):
+    my_conn.request("/update")
+
+
+@job
+def update_service_job():
+    update_service()
+
+
+defs = Definitions(
+    jobs=[update_service_job],
+    resources={
+        "my_conn": MyConnectionResource(username="my_user", password="my_password"),
+    },
+)
+```
+
+</TabItem>
+</TabGroup>
+
+---
+
+## Using environment variables with resources
+
+Resources can be configured using environment variables, which is useful for secrets or other environment-specific configuration.
+
+To use environment variables, pass an <PyObject object="EnvVar" /> when constructing your resource.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_env_vars endbefore=end_new_resources_env_vars
+from dagster._config.field_utils import EnvVar
+from dagster import Definitions
+from dagster._config.structured_config import ConfigurableResource
+
+
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+defs = Definitions(
+    assets=...,
+    resources={
+        "credentials": CredentialsResource(
+            username=EnvVar("MY_USERNAME"),
+            password=EnvVar("MY_PASSWORD"),
+        )
+    },
+)
+```
+
+---
+
+## Configuring resources at runtime
+
+In some cases, you may want to specify configuration for a resource at runtime, in the launchpad. For example, you may want to specify a different target table in a database resource for each run.
+
+To do this, you can use the `configure_at_launch()` method to defer the construction of a configurable resource until runtime.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime endbefore=end_new_resource_runtime
+class DatabaseResource(ConfigurableResource):
+    table: str
+
+    def read(self):
+        ...
+
+
+@asset
+def data_from_database(db_conn: DatabaseResource):
+    return db_conn.read()
+
+
+defs = Definitions(
+    assets=...,
+    resources={"db_conn": DatabaseResource.configure_at_launch()},
+)
+```
+
+---
+
+## Resources which depend on other resources
+
+In some situations, you may want to define a resource which depends on other resources. For example, a resource representing a connection to an external service may rely on a resource representing credentials for that service.
```

Comment:
I think we need to flesh out the motivation here.

Reply:
added some more explanation